File: assign4/ruby/lib/guess.rb
def guess(selection, attempts, input, output)        
  plural = if attempts == 1 then "" else "s" end
  responses = ["Aim lower", "You got it in #{attempts} attempt#{plural}", "Aim higher"]

  guess_match = (selection - input.call) <=> 0
  output.call(responses[guess_match + 1]) 
  
  guess_match == 0    
end

def play(selection, input, output)
  (1..).find { |attempts| guess(selection, attempts, input, output) } 
end

File: assign4/julia/assign/src/guess.jl
module guessNumber

  using Base.Iterators
  using Lazy
  using IterTools

  export guess, play

  function guess(selection, attempts, input, output)
      plural = attempts > 1 ? "s" : ""
      responses = ["Aim lower", "You got it in $attempts attempt$plural", "Aim higher"]
      guess_match = sign(selection - input())
      output(responses[guess_match + 2])
      guess_match == 0
  end

  function play(selection, input, output)
      collect(IterTools.takewhile(attempt -> !guess(selection, attempt, input, output), Lazy.range(1)))
  end
end
